### ISSUE TITLE
unlink of log file that no longer exists should be handled as a warning only

### ISSUE THREAD

**Issue by gp4git** [logrotate/logrotate#144]
Hello,
*** For the record, it seems very similar with issue #110 ***
Logrotate version : 3.8.6
Config file :
```
/var/log/pgsql96/pgsql96.log {
  compress
  delaycompress
  notifempty
  size 50M
  daily
  rotate 31
  create 0640 postgres postgres
  postrotate
  /bin/kill -HUP `cat /var/run/syslogd.pid 2> /dev/null` 2> /dev/null || true
  find /var/log/pgsql96/* -mtime +30 -exec rm -f  {} \; 2>&1 > /dev/null
  endscript
}
```
Here is the output in verbose mode, I got last night :
```
rotating pattern: /var/log/pgsql96/pgsql96.log after 1 days (31 rotations)
empty log files are not rotated, old logs are removed
considering log /var/log/pgsql96/pgsql96.log
log needs rotating
rotating log /var/log/pgsql96/pgsql96.log, log->rotateCount is 31
dateext suffix '-20170914'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
compressing log with: /bin/gzip
fscreate context set to unconfined_u:object_r:var_log_t:s0
renaming /var/log/pgsql96/pgsql96.log to /var/log/pgsql96/pgsql96.log-20170914
creating new /var/log/pgsql96/pgsql96.log mode = 0640 uid = 26 gid = 26
running postrotate script
removing old log /var/log/pgsql96/pgsql96.log-20170814.gz
error: error opening /var/log/pgsql96/pgsql96.log-20170814.gz: No such file or directory
```

Now, after pasting the conf we use, I think it may be related to postrotate directive find ...-mtime +30
but logrotate should ignore this missing file error.

Regards

**Comment by kdudka** [comment]
So the issue is that logrotate tries to remove a rotated file that has already been removed by a script.  In your case it fails on an unnecessary call to ```open()``` which was eliminated in 0d21e082 but even with the latest code it would fail on ```unlink()```.  We could probably lower the severity to warning if ```unlink()``` fails with ```ENOENT```?

**Comment by gp4git** [comment]
I must add that this exact same configuration  works without error on server with logrotate version 3.7.8

**Comment by kdudka** [comment]
I do not think it is related to logrotate version.  It just depends whether old log files are removed by the script before logrotate itself tries to remove them (which itself depends on many factors in your case).

**Comment by gp4git** [comment]
To be complete on our conf, we also  have dateext in /etc/logrotate.conf
Interactively, I try to reproduce the issue but with no success till now.
I cannot get the message "removing old log.." in verbose or debug mode.
Any idea when logrotate code emit this message ?

Regards

**Comment by kdudka** [comment]
minimal example:
```
% cat > /tmp/lr.conf << EOF
/tmp/logs/logfile {
    create
    rotate 1
    size 0
    postrotate
        rm -fv /tmp/logs/logfile.2
    endscript
}
EOF

% mkdir /tmp/logs
% touch /tmp/logs/logfile{,.1}
% logrotate -s /tmp/lr.status -v /tmp/lr.conf
reading config file /tmp/lr.conf
Reading state from file: /tmp/lr.status
Allocating hash table for state file, size 64 entries

Handling 1 logs

rotating pattern: /tmp/logs/logfile  0 bytes (1 rotations)
empty log files are rotated, old logs are removed
considering log /tmp/logs/logfile
Creating new state
  Now: 2017-09-15 16:32
  Last rotated at 2017-09-15 16:00
  log needs rotating
rotating log /tmp/logs/logfile, log->rotateCount is 1
dateext suffix '-20170915'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/logs/logfile.1 to /tmp/logs/logfile.2 (rotatecount 1, logstart 1, i 1), 
renaming /tmp/logs/logfile.0 to /tmp/logs/logfile.1 (rotatecount 1, logstart 1, i 0), 
old log /tmp/logs/logfile.0 does not exist
renaming /tmp/logs/logfile to /tmp/logs/logfile.1
creating new /tmp/logs/logfile mode = 0644 uid = 1024 gid = 1024
running postrotate script
removed '/tmp/logs/logfile.2'
removing old log /tmp/logs/logfile.2
error: Failed to remove old log /tmp/logs/logfile.2: No such file or directory

% echo $?
1
```

**Comment by gp4git** [comment]
OK, so it happens only when the old log is removed INSIDE the postrotate script and not manually outside the logrotate script.
I asked internally why the "find -rm" command  was included in postrotate conf but I get no answer for the moment.
Maybe it would be safer to use "maxage" in this case. What do you think ?

**Comment by kdudka** [comment]
Definitely.  If you look at the examples in logrotate man page, you will see that ```postrotate``` scripts are mostly used for sending ```SIGHUP``` signals to daemons after rotation to make them start using a newly created log file.  It is definitely not intended for removing aged log files (unlike ```maxage``` as you suggest).

On the other hand, lowering the severity of ```unlink()``` failing with ```ENOENT``` from error to warning should be both safe and useful.

**Comment by gp4git** [comment]
Agreed for lowering from error to warning.
Thank you for your time.
I close the ticket.
Regards

**Comment by kdudka** [comment]
No problem.  Reopening the ticket until the change is actually pushed...

**Comment by marXtevens** [comment]
Fast forward to the end of 2020. I have noticed a re-occurrence of this problem. 
It does not happen every rotation. I am guessing it is a timing concern, and still generates an error, rather than a warning?
The result is that logrotated does not restart automatically.

I am running Fedora 32 Server, and updating weekly. 
I am attaching output from journalctl for logrotate for the beginning of November, through December 3 (today). 
[journalctl_logrotate.zip](https://github.com/logrotate/logrotate/files/5637313/journalctl_logrotate.zip)

**Comment by kdudka** [comment]
The presentation of (fatal) errors vs. warnings in logrotate is a mess.  We have a separate pull request to clean it up: #239.  The fix for this issue (commit 123557b36eda5fe08b1f2f8b8c7644df33683147) just makes logrotate continue and return success exit status in case `unlink()` fails.  But it still prints the error, which is in fact a warning only.  I think the problems you are observing must be unrelated to the error/warning in question.  We can of course change the message to make it clear that it is a warning only.  If you think it would helped, please file it as a separate issue or pull request.
